docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+The deduplication period chosen by the ledger is the *effective deduplication period*.
+The ledger may also convert a requested deduplication duration into an effective deduplication offset or vice versa.
+The effective deduplication period is reported in the the command completion event in the :ref:`deduplication duration <com.daml.ledger.api.v1.Completion.deduplication_duration>` or :ref:`deduplication offset <com.daml.ledger.api.v1.Completion.deduplication_offset>` fields.
+
+A command submission is considered a **duplicate submission** if at least one of the following holds:
+
+- The submitting participant's completion service contains a successful completion event for the same :ref:`change ID <change-id>` within the *effective* deduplication period.
+
+- The participant or Daml ledger are aware of another command submission in-flight with the same :ref:`change ID <change-id>` when they performs command deduplication.
+
+Command submissions via the :ref:`command service <command-service>` indicate the command deduplication outcome as a synchronous gRPC response (except when the gRPC deadline was exceeded before).
+Submissions via the :ref:`command submission service <command-submission-service>` can indicate the outcome synchronously or asynchronously in the event on the :ref:`command completion service <command-completion-service>`.
+In particular, the submission may be a duplicate even if the command submission service acknowledges the submission with the gRPC status code ``OK``.
+Command deduplication generates the following outcomes of a command submission:
+
+- If no conflicting submission with the same :ref:`change ID <change-id>` was found, the completion event and the response convey the result of the submission (success or a gRPC error).
+
+- The gRPC status code ``ALREADY_EXISTS`` with error code id :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>` if there is an earlier command completion for the same :ref:`change ID <change-id>` within the effective deduplication period.
+
+- The gRPC status code ``ABORTED`` with error code id :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` if another submission for the same :ref:`change ID <change-id>` was in flight when this submission was processed.
+
+- The gRPC status code ``FAILED_PRECONDITION`` with error code id :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` if the specified deduplication period is not supported.
+  The fields ``longest_duration`` or ``earliest_offset`` in the metadata specify the longest duration or earliest offset that is currently supported on the Ledger API endpoint.
+  At least one of the two fields is present.
+
+  Deduplication periods that span at most the :ref:`maximum deduplication time <com.daml.ledger.api.v1.LedgerConfiguration.max_deduplication_time>` SHOULD not result in this error.
+  Participants may accept longer periods at their discretion.
```

Comment:
I agree. The current documentation makes it sound like the maximum is useless:

1. Users can supply periods longer than the "maximum"
2. The participant can extend periods beyond the "maximum"

I assume (2) is because of technical reasons the participant/ledger may have to extend the period by a small duration.
What is the reason for (1)? Validating the duration is a bit annoying because the participant doesn't know the latest ledger configuration and there could be race conditions. However, the ledger should be able to validate it the same way it validates the ledger time model (which also depends both on the current configuration and the submitted transaction)

Reply:
I agree that `maximum` is a misleading name here. I'm happy to go with a different name. The reason why it's still called `maximum deduplication time` is because that's the name of the field in the ledger configuration protobuf message and we didn't dare to rename it (yet).

Reply:
IIUC it's really the "maximum deduplication time guaranteed to be supported by the ledger / operator" but that's very verbose and I am not sure how it could be shortened. Also, a naming misalignment with the proto would indeed be a problem.

Reply:
> I assume (2) is because of technical reasons the participant/ledger may have to extend the period by a small duration.

Not only for technical reasons. If the ledger reports a longer dedup period, then the application gains more knowledge. So longer dedup periods are in a sense better.

> What is the reason for (1)?

The reasoning is as follows: If it is no effort for the participant / ledger to support longer dedup period, then this is something we should make available to applications on a best-effort bases. They shouldn't rely on it in general. But suppose we are in the unforeseen eventuality that an outage has lasted for long enough that the max deduplication time is too short (especially when a deduplication offset is used). Then it just feels wrong to say "sorry, I can't help you with this because your deduplication offset is too early" even though it can be supported without problems because the participant happens to still have the data available.

Reply:
> maximum deduplication time guaranteed to be supported by the ledger / operator

I like this explanation (and also the longer explanation by @andreaslochbihler-da helped clearing it up for me).